spring-cloud-stream-binders/spring-cloud-stream-binder-rabbit/src/main/java/org/springframework/cloud/stream/binder/rabbit/RabbitMessageChannelBinder.java

```diff
@@ -587,54 +521,36 @@ private MessageRecoverer determineRecoverer(String name, RabbitPropertiesAccesso
 		}
 	}
 
-	@Override
-	public void bindProducer(final String name, MessageChannel moduleOutputChannel,
-			Properties properties) {
-		Assert.isInstanceOf(SubscribableChannel.class, moduleOutputChannel);
-		RabbitPropertiesAccessor accessor = new RabbitPropertiesAccessor(properties);
-		if (name.startsWith(P2P_NAMED_CHANNEL_TYPE_PREFIX)) {
-			validateProducerProperties(name, properties, SUPPORTED_NAMED_PRODUCER_PROPERTIES);
-		}
-		else {
-			validateProducerProperties(name, properties, SUPPORTED_PRODUCER_PROPERTIES);
-		}
-		if (!bindNewProducerDirectlyIfPossible(name, (SubscribableChannel) moduleOutputChannel, accessor)) {
-			if (this.logger.isInfoEnabled()) {
-				this.logger.info("declaring queue for outbound: " + name);
-			}
-			AmqpOutboundEndpoint queue = this.buildOutboundEndpoint(name, accessor, determineRabbitTemplate(accessor));
-			doRegisterProducer(name, moduleOutputChannel, queue, accessor);
-		}
-	}
-
 	private AmqpOutboundEndpoint buildOutboundEndpoint(final String name, RabbitPropertiesAccessor properties,
 			RabbitTemplate rabbitTemplate) {
 		String prefix = properties.getPrefix(this.defaultPrefix);
-		String queueName = applyPrefix(prefix, name);
+		String exchangeName = applyPrefix(prefix, name);
 		String partitionKeyExtractorClass = properties.getPartitionKeyExtractorClass();
 		Expression partitionKeyExpression = properties.getPartitionKeyExpression();
-		TopicExchange exchange = new TopicExchange(queueName);
-		declareExchange(queueName, exchange);
+		TopicExchange exchange = new TopicExchange(exchangeName);
+		declareExchange(exchangeName, exchange);
 		AmqpOutboundEndpoint endpoint = new AmqpOutboundEndpoint(rabbitTemplate);
 		endpoint.setExchangeName(exchange.getName());
+		String baseQueueName = exchangeName + ".default";
 		if (partitionKeyExpression == null && !StringUtils.hasText(partitionKeyExtractorClass)) {
-			Queue queue = new Queue(queueName, true, false, false, queueArgs(properties, queueName));
-			declareQueue(name, queue);
-			autoBindDLQ(name, properties);
-			endpoint.setRoutingKey(queueName);
-			org.springframework.amqp.core.Binding binding = BindingBuilder.bind(queue).to(exchange).with(queueName);
-			declareBinding(queueName, binding);
+			Queue queue = new Queue(baseQueueName, true, false, false, queueArgs(properties, baseQueueName));
+			declareQueue(baseQueueName, queue);
+			autoBindDLQ(baseQueueName, baseQueueName, properties);
+			endpoint.setRoutingKey(name);
+			org.springframework.amqp.core.Binding binding = BindingBuilder.bind(queue).to(exchange).with(name);
```

Comment:
My reason for routing with the full queue name is to make DLQ re-queueing easier.

We route to the DLX using the full queue name (we set it explicitly on the queue but that's not really necessary because rabbit will route to the DLX using the original routing key if none is specified on the queue).

If we route to the primary queue using just the partition, we'd still need to route to the DLX using the full queue name.

If I want to requeue a dead lettered message, I can simply route to the proper exchange using the message's `receivedRoutingKey`.

If we change the primary routing key we'd need to manipulate the received routing key when re-queuing - or use a separate DLX for each binding. (Currently all DLQs are bound to a single DLX).

That said, the `x-death` header has all the information we need to re-route (exchange, routing key), which we'd need to examine to get the exchange anyhow.

But yes, we could bind the default queue with a hash.